utilities/src/lib.rs

```diff
@@ -0,0 +1,349 @@
+        zip_first.read_to_string(&mut zip_first_contents).expect("Failed to read zip file content");
+        fs::write(path_dat, zip_first_contents).expect("Failed to write zip content to dat file");
+    }
+    Ok(())
+}
+
+pub fn as_f64(s: &str) -> Result<f64, Box<dyn Error + Sync + Send>> {
+    match s {
+        "nan" => Ok(f64::NAN),
+        "-nan" => Ok(-f64::NAN),
+        "inf" => Ok(f64::INFINITY),
+        "-inf" => Ok(f64::NEG_INFINITY),
+        _ => {
+            if s.starts_with("0x") {
+                // Bit fields are recorded using hex notation, to avoid endian difference complications.
+                let without_prefix = s.trim_start_matches("0x");
+                match u64::from_str_radix(without_prefix, 16) {
+                    Ok(num) => Ok(num as f64),
+                    Err(err) => Err(Box::<dyn Error + Sync + Send>::from(err.to_string())),
+                }
+            } else {
+                // All other value types are recorded using values that can be parsed as f64.
+                // In some cases, they'll need to be coerced into a different type later.
+                match s.parse::<f64>() {
+                    Ok(num) => Ok(num),
+                    Err(err) => Err(Box::<dyn Error + Sync + Send>::from(err.to_string())),
+                }
+            }
+        },
+    }
+}
+
+// Return the absolute and relative difference between two values, in that order.
+// If both values are nan, consider the difference to be 0.
+pub fn calc_delta(x: f64, y: f64) -> (f64, f64) {
+    if x.is_nan() && y.is_nan() {
+        return (0f64, 0f64);
+    } else if x.is_infinite() && y.is_infinite() {
+        return if x.is_sign_negative() == y.is_sign_negative() { (0f64, 0f64) } else { (f64::INFINITY, f64::INFINITY) };
+    } else {
+        let diff_abs = (x - y).abs();
+        let diff_rel = 2.0 * diff_abs / (x.abs() + y.abs());
+        return (diff_abs, diff_rel);
+    }
+}
+
+// Extract a the first value from each line, and return as a vector.
+pub fn as_vec_basic(op_name: &str) -> Vec<f64> {
+    let paths = get_data_paths(op_name).expect("Failed to determine zip and dat paths");
+    refresh_unzipped(&paths).expect("Failed to refresh unzipped data file");
+    let file = File::open(paths.path_dat.as_path()).unwrap();
+    let reader = io::BufReader::new(file);
+    let result: Vec<f64> = reader.lines().enumerate()
+        // Skip the header line
+        .filter(|(i, _line)| *i != 0)
+        .map(|(_i, line)| {
+            let line_safe = line.expect("Failed to read line");
+            let item = line_safe.splitn(2, ' ').nth(0).unwrap();
+            as_f64(item).expect("Failed to parse item")
+        })
+        .collect();
+        result
+}
+
+// We can't use f64 in the key directly because it implements PartialEq,
+// but not Eq or Hash, because of the complication that NaN != NaN.
+pub fn get_geodesic_lookup(data: &Vec<Vec<f64>>) -> HashMap<(u64, u64), Geodesic> {
```

Comment:
Sorry, I don't see where this is used

Reply:
It's used in many geodesic_benchmark functions, including this example snippet:

```rust
fn benchmark_geodesic_gen_direct(c: &mut Criterion) {
    // Format: this-in[_a _f] lat1 lon1 azi1 arcmode s12_a12 outmask result=a12 lat2-out lon2-out azi2-out s12-out m12-out M12-out M21-out S12-out
    let data = utilities::as_vecs_basic("Geodesic_GenDirect", 8);
    let geods = utilities::get_geodesic_lookup(&data);
```

As I mentioned in my previous comment, each data line starts with the operation's inputs. That also includes any relevant state information. For operations in the context of a geodesic instance, the first two values are typically the `a` (equatorial radius) and `f` (flattening) values of the geodesic. This is enough information to unambiguously identify or reconstruct the geodesic. We could simply construct a fresh geodesic for every data line in a benchmark, but that would result in spending a lot of time on something that isn't the operation of interest to the benchmark. Instead, I have such benchmarks start by making a sweep through the data in non-timed code and construct each unique geodesic, and store it in a lookup keyed by the `a` and `f` values. I think it usually ends up being 20-ish geodesics. During the timed portion, it can then just get the appropriate geodesic from the lookup, which is much cheaper than constructing a geodesic.